app/Policies/AlbumPolicy.php

```diff
@@ -0,0 +1,270 @@
+<?php
+
+namespace App\Policies;
+
+use App\Contracts\AbstractAlbum;
+use App\Exceptions\ConfigurationKeyMissingException;
+use App\Exceptions\Internal\FrameworkException;
+use App\Exceptions\Internal\LycheeAssertionError;
+use App\Exceptions\Internal\QueryBuilderException;
+use App\Factories\AlbumFactory;
+use App\Models\BaseAlbumImpl;
+use App\Models\Configs;
+use App\Models\Extensions\BaseAlbum;
+use App\Models\User;
+use App\SmartAlbums\BaseSmartAlbum;
+use Illuminate\Auth\Access\HandlesAuthorization;
+use Illuminate\Contracts\Container\BindingResolutionException;
+use Illuminate\Support\Facades\Session;
+
+class AlbumPolicy
+{
+	use HandlesAuthorization;
+
+	protected UserPolicy $userPolicy;
+
+	public const UNLOCKED_ALBUMS_SESSION_KEY = 'unlocked_albums';
+
+	// constants to be used in GATE
+	public const OWN = 'own';
+	public const ACCESS = 'access';
+	public const DOWNLOAD = 'download';
+	public const EDIT = 'edit';
+	public const VISIBLE = 'visible';
+
+	/**
+	 * @throws FrameworkException
+	 */
+	public function __construct()
+	{
+		try {
+			$this->userPolicy = resolve(UserPolicy::class);
+		} catch (BindingResolutionException $e) {
+			throw new FrameworkException('Laravel\'s provider component', $e);
+		}
+	}
+
+	/**
+	 * Perform pre-authorization checks.
+	 *
+	 * @param \App\Models\User $user
+	 * @param string           $ability
+	 *
+	 * @return void|bool
+	 */
+	public function before(?User $user, $ability)
+	{
+		if ($this->userPolicy->admin($user)) {
+			return true;
+		}
+	}
+
+	/**
+	 * This gate policy ensures that the Album is owned by current user.
+	 * Do note that in case of current user being admin, it will be skipped due to the before method.
+	 *
+	 * @param User|null $user
+	 * @param BaseAlbum $album
+	 *
+	 * @return bool
+	 */
+	public function own(?User $user, BaseAlbum $album): bool
```

Comment:
I know that the example in the official Laravel documentation is written this way, but personally I don't like the pattern how the method names are chosen. Everywhere else we follow the paradigm that methods start with a verb which expresses what the method does and that methods which return a boolean either start with `is` or `has`.

This would imply the following changes (examples only)

- `own` --> `isOwner` or `hasOwnership`
- `edit`--> `isEditable` or at least `canEdit`
- `visible` --> `isVisibile` ("visibile" alone is not even a verb)
- `download` --> `isDownloadable` or at least `canDownload`

In particular `edit` and `download` bother me the most (and some other not mentioned here). These method names are indeed verbs, but the methods do _not_ edit the given album or download the given album, but simply check if it is OK to do so.

Personally, I don't like this style and it appears confusing to me, because I cannot intuitively grasp from the name of the method what the method does (or what it does not).

@LycheeOrg/reviewers Opinions?

Reply:
renamed.